rollup/core/classes/RollupState.cls

```diff
@@ -1,60 +1,111 @@
-public without sharing virtual class RollupState {
-  private final Map<String, Map<Type, RollupState>> keyToState {
+public without sharing virtual class RollupState implements System.Queueable, System.Finalizer, RollupLogger.ToStringObject {
+  protected transient String key;
+  protected transient Integer keyLength;
+  protected transient String recordId;
+  protected transient String typeName;
+  protected final Set<String> jobIds = new Set<String>();
+
+  private Long commitCount = 1;
+  private Set<Id> statefulPreviouslyRetrievedStateIds = new Set<Id>();
+
+  @TestVisible
+  private static Integer maxBodyLength = 131072;
+  @TestVisible
+  private static Integer maxRelatedKeysLength = 255;
+
+  private static final Map<String, RollupState> KEY_TO_STATE {
     get {
-      this.keyToState = this.keyToState ?? new Map<String, Map<Type, RollupState>>();
-      return this.keyToState;
+      KEY_TO_STATE = KEY_TO_STATE ?? new Map<String, RollupState>();
+      return KEY_TO_STATE;
     }
     set;
   }
 
-  public RollupState getState(String firstKey, Type secondKey) {
-    return this.keyToState.get(firstKey)?.get(secondKey);
+  private static final Map<String, List<RollupState__c>> CACHED_STATES {
+    get {
+      CACHED_STATES = CACHED_STATES ?? new Map<String, List<RollupState__c>>();
+      return CACHED_STATES;
+    }
+    set;
   }
 
-  public virtual Type getType() {
-    return RollupState.class;
+  public RollupState getState(String key, Rollup__mdt meta, Type initializingType) {
+    String trueKey = (key + getMetadataKey(meta));
+    RollupState possibleState = KEY_TO_STATE.get(trueKey);
+    if (possibleState == null) {
+      possibleState = (RollupState) initializingType.newInstance();
+      possibleState.recordId = key;
+      this.setState(trueKey, key.length(), possibleState);
```

Comment:
I can't decide, is it worth adding a comment somewhere in here about `trueKey` vs `key`, and `key.length()`? At first, I thought that since you're passing `trueKey` here, it should use `trueKey.length()` - I figured out that `key.length()` does make sense here, but I had to hop around a little in the class to confirm that... but I'm still not sure that an extra comment is needed 🤷